File: go/config_test.go
package haisdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

func TestLoadConfig(t *testing.T) {
	tmpDir := t.TempDir()
	configPath := filepath.Join(tmpDir, "jacs.config.json")

	cfg := Config{
		JacsAgentName:    "test-agent",
		JacsAgentVersion: "1.0.0",
		JacsKeyDir:       "/keys",
		JacsID:           "jacs-id-123",
	}
	data, _ := json.Marshal(cfg)
	if err := os.WriteFile(configPath, data, 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	loaded, err := LoadConfig(configPath)
	if err != nil {
		t.Fatalf("LoadConfig: %v", err)
	}

	if loaded.JacsAgentName != "test-agent" {
		t.Errorf("expected JacsAgentName 'test-agent', got '%s'", loaded.JacsAgentName)
	}
	if loaded.JacsID != "jacs-id-123" {
		t.Errorf("expected JacsID 'jacs-id-123', got '%s'", loaded.JacsID)
	}
	if loaded.JacsKeyDir != "/keys" {
		t.Errorf("expected JacsKeyDir '/keys', got '%s'", loaded.JacsKeyDir)
	}
}

func TestLoadConfigNotFound(t *testing.T) {
	_, err := LoadConfig("/nonexistent/jacs.config.json")
	if err == nil {
		t.Fatal("expected error for missing config")
	}

	sdkErr, ok := err.(*Error)
	if !ok {
		t.Fatalf("expected *Error, got %T", err)
	}
	if sdkErr.Kind != ErrConfigNotFound {
		t.Errorf("expected ErrConfigNotFound, got %v", sdkErr.Kind)
	}
}

func TestLoadConfigInvalidJSON(t *testing.T) {
	tmpDir := t.TempDir()
	configPath := filepath.Join(tmpDir, "jacs.config.json")
	if err := os.WriteFile(configPath, []byte("not json{{{"), 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	_, err := LoadConfig(configPath)
	if err == nil {
		t.Fatal("expected error for invalid JSON")
	}

	sdkErr, ok := err.(*Error)
	if !ok {
		t.Fatalf("expected *Error, got %T", err)
	}
	if sdkErr.Kind != ErrConfigInvalid {
		t.Errorf("expected ErrConfigInvalid, got %v", sdkErr.Kind)
	}
}

func TestDiscoverConfigFromEnv(t *testing.T) {
	tmpDir := t.TempDir()
	configPath := filepath.Join(tmpDir, "jacs.config.json")

	cfg := Config{
		JacsAgentName: "env-agent",
		JacsID:        "env-id",
	}
	data, _ := json.Marshal(cfg)
	_ = os.WriteFile(configPath, data, 0o644)

	t.Setenv("JACS_CONFIG_PATH", configPath)

	loaded, err := DiscoverConfig()
	if err != nil {
		t.Fatalf("DiscoverConfig: %v", err)
	}
	if loaded.JacsAgentName != "env-agent" {
		t.Errorf("expected 'env-agent', got '%s'", loaded.JacsAgentName)
	}
}

func TestDiscoverConfigNotFound(t *testing.T) {
	// Clear env and ensure no config in cwd or home
	t.Setenv("JACS_CONFIG_PATH", "")

	// Change to a temp dir with no config
	tmpDir := t.TempDir()
	origDir, _ := os.Getwd()
	_ = os.Chdir(tmpDir)
	defer func() { _ = os.Chdir(origDir) }()

	_, err := DiscoverConfig()
	if err == nil {
		t.Fatal("expected error when no config exists")
	}
}

func TestResolveKeyPath(t *testing.T) {
	cfg := &Config{
		JacsAgentName: "my-agent",
		JacsKeyDir:    "/custom/keys",
	}

	path := ResolveKeyPath(cfg, "/some/dir/jacs.config.json")
	expected := "/custom/keys/agent_private_key.pem"
	if path != expected {
		t.Errorf("expected '%s', got '%s'", expected, path)
	}
}

func TestResolveKeyPathDefaultDir(t *testing.T) {
	cfg := &Config{
		JacsAgentName: "my-agent",
		JacsKeyDir:    "", // empty -> use config dir
	}

	path := ResolveKeyPath(cfg, "/home/user/.jacs/jacs.config.json")
	expected := "/home/user/.jacs/agent_private_key.pem"
	if path != expected {
		t.Errorf("expected '%s', got '%s'", expected, path)
	}
}

func TestResolveKeyPathRelativeDirUsesConfigLocation(t *testing.T) {
	cfg := &Config{
		JacsAgentName: "my-agent",
		JacsKeyDir:    "keys",
	}

	path := ResolveKeyPath(cfg, "/home/user/.jacs/jacs.config.json")
	expected := "/home/user/.jacs/keys/agent_private_key.pem"
	if path != expected {
		t.Errorf("expected '%s', got '%s'", expected, path)
	}
}

func TestResolvePrivateKeyPasswordFromEnv(t *testing.T) {
	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", "dev-password")
	t.Setenv("JACS_PASSWORD_FILE", "")

	pwd, err := ResolvePrivateKeyPassword()
	if err != nil {
		t.Fatalf("ResolvePrivateKeyPassword: %v", err)
	}
	if string(pwd) != "dev-password" {
		t.Fatalf("unexpected password: %q", string(pwd))
	}
}

func TestResolvePrivateKeyPasswordFromFileWhenEnvDisabled(t *testing.T) {
	tmpDir := t.TempDir()
	passwordFile := filepath.Join(tmpDir, "password.txt")
	if err := os.WriteFile(passwordFile, []byte("file-password\n"), 0o600); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", "dev-password")
	t.Setenv("JACS_PASSWORD_FILE", passwordFile)
	t.Setenv("JACS_DISABLE_PASSWORD_ENV", "1")

	pwd, err := ResolvePrivateKeyPassword()
	if err != nil {
		t.Fatalf("ResolvePrivateKeyPassword: %v", err)
	}
	if string(pwd) != "file-password" {
		t.Fatalf("unexpected password: %q", string(pwd))
	}
}

func TestResolvePrivateKeyPasswordFailsOnMultipleSources(t *testing.T) {
	tmpDir := t.TempDir()
	passwordFile := filepath.Join(tmpDir, "password.txt")
	if err := os.WriteFile(passwordFile, []byte("file-password\n"), 0o600); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", "dev-password")
	t.Setenv("JACS_PASSWORD_FILE", passwordFile)

	_, err := ResolvePrivateKeyPassword()
	if err == nil {
		t.Fatal("expected conflict error for multiple password sources")
	}
}

func TestResolvePrivateKeyPasswordFailsWhenMissing(t *testing.T) {
	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", "")
	t.Setenv("JACS_PASSWORD_FILE", "")

	_, err := ResolvePrivateKeyPassword()
	if err == nil {
		t.Fatal("expected missing password source error")
	}
}

func TestResolvePrivateKeyPasswordFailsOnInsecureFilePermissions(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("permission-mode checks are unix-specific")
	}

	tmpDir := t.TempDir()
	passwordFile := filepath.Join(tmpDir, "password.txt")
	if err := os.WriteFile(passwordFile, []byte("file-password\n"), 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}
	if err := os.Chmod(passwordFile, 0o644); err != nil {
		t.Fatalf("Chmod: %v", err)
	}

	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", "")
	t.Setenv("JACS_PASSWORD_FILE", passwordFile)
	t.Setenv("JACS_DISABLE_PASSWORD_ENV", "1")

	_, err := ResolvePrivateKeyPassword()
	if err == nil {
		t.Fatal("expected insecure permissions error")
	}
	if !strings.Contains(err.Error(), "insecure permissions") {
		t.Fatalf("expected insecure permissions error, got: %v", err)
	}
}

func TestResolvePrivateKeyPasswordFailsOnSymlinkFile(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("symlink checks are unix-specific in CI/dev")
	}

	tmpDir := t.TempDir()
	targetFile := filepath.Join(tmpDir, "password-target.txt")
	linkFile := filepath.Join(tmpDir, "password-link.txt")
	if err := os.WriteFile(targetFile, []byte("file-password\n"), 0o600); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}
	if err := os.Chmod(targetFile, 0o600); err != nil {
		t.Fatalf("Chmod: %v", err)
	}
	if err := os.Symlink(targetFile, linkFile); err != nil {
		t.Fatalf("Symlink: %v", err)
	}

	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", "")
	t.Setenv("JACS_PASSWORD_FILE", linkFile)
	t.Setenv("JACS_DISABLE_PASSWORD_ENV", "1")

	_, err := ResolvePrivateKeyPassword()
	if err == nil {
		t.Fatal("expected symlink rejection error")
	}
	if !strings.Contains(err.Error(), "must not be a symlink") {
		t.Fatalf("expected symlink rejection error, got: %v", err)
	}
}

File: node/tests/config.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { loadConfig, loadPrivateKey, loadPrivateKeyPassphrase } from '../src/config.js';
import { generateTestKeypair as generateKeypair } from './setup.js';
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import * as os from 'node:os';

describe('config', () => {
  let tmpDir: string;

  beforeEach(async () => {
    tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'haiai-test-'));
  });

  afterEach(async () => {
    delete process.env.JACS_CONFIG_PATH;
    delete process.env.JACS_PRIVATE_KEY_PASSWORD;
    delete process.env.JACS_PASSWORD_FILE;
    delete process.env.JACS_DISABLE_PASSWORD_ENV;
    delete process.env.JACS_DISABLE_PASSWORD_FILE;
    await fs.rm(tmpDir, { recursive: true, force: true });
  });

  describe('loadConfig', () => {
    it('loads config from explicit path', async () => {
      const configPath = path.join(tmpDir, 'jacs.config.json');
      await fs.writeFile(configPath, JSON.stringify({
        jacsAgentName: 'test-bot',
        jacsAgentVersion: '2.0.0',
        jacsKeyDir: './keys',
        jacsId: 'jacs-id-1',
      }));

      const config = await loadConfig(configPath);
      expect(config.jacsAgentName).toBe('test-bot');
      expect(config.jacsAgentVersion).toBe('2.0.0');
      expect(config.jacsKeyDir).toBe(path.resolve(tmpDir, 'keys'));
      expect(config.jacsId).toBe('jacs-id-1');
    });

    it('loads config with canonical snake_case fields', async () => {
      const configPath = path.join(tmpDir, 'jacs.config.json');
      await fs.writeFile(configPath, JSON.stringify({
        jacs_agent_id_and_version: 'snake-id:1.0.0',
        jacs_key_directory: './mykeys',
      }));

      const config = await loadConfig(configPath);
      expect(config.jacsAgentVersion).toBe('1.0.0');
      expect(config.jacsKeyDir).toBe(path.resolve(tmpDir, 'mykeys'));
      expect(config.jacsId).toBe('snake-id');
    });

    it('throws when required fields are missing', async () => {
      const configPath = path.join(tmpDir, 'jacs.config.json');
      await fs.writeFile(configPath, JSON.stringify({}));

      await expect(loadConfig(configPath)).rejects.toThrow(
        /JACS config has neither canonical nor legacy fields/,
      );
    });

    it('throws for non-existent file', async () => {
      await expect(loadConfig('/non/existent/path.json')).rejects.toThrow();
    });

    it('throws for invalid JSON', async () => {
      const configPath = path.join(tmpDir, 'bad.json');
      await fs.writeFile(configPath, 'not-json');
      await expect(loadConfig(configPath)).rejects.toThrow();
    });

    it('uses JACS_CONFIG_PATH env var', async () => {
      const configPath = path.join(tmpDir, 'jacs.config.json');
      await fs.writeFile(configPath, JSON.stringify({
        jacsAgentName: 'env-bot',
        jacsAgentVersion: '1.0.0',
        jacsKeyDir: '.',
      }));

      process.env.JACS_CONFIG_PATH = configPath;
      const config = await loadConfig();
      expect(config.jacsAgentName).toBe('env-bot');
      expect(config.jacsKeyDir).toBe(path.resolve(tmpDir));
    });

    it('resolves relative jacsPrivateKeyPath from config directory', async () => {
      const configPath = path.join(tmpDir, 'jacs.config.json');
      await fs.writeFile(configPath, JSON.stringify({
        jacsAgentName: 'test-bot',
        jacsAgentVersion: '1.0.0',
        jacsKeyDir: './keys',
        jacsPrivateKeyPath: './custom/agent_private_key.pem',
      }));

      const config = await loadConfig(configPath);
      expect(config.jacsPrivateKeyPath).toBe(
        path.resolve(tmpDir, 'custom/agent_private_key.pem'),
      );
    });
  });

  describe('loadPrivateKey', () => {
    it('loads private key from key directory', async () => {
      const keyDir = path.join(tmpDir, 'keys');
      await fs.mkdir(keyDir);

      const kp = generateKeypair();
      await fs.writeFile(path.join(keyDir, 'private_key.pem'), kp.privateKeyPem);

      const config = {
        jacsAgentName: 'test',
        jacsAgentVersion: '1.0',
        jacsKeyDir: keyDir,
      };

      const pem = await loadPrivateKey(config);
      expect(pem).toContain('-----BEGIN PRIVATE KEY-----');
    });

    it('loads from jacsPrivateKeyPath if specified', async () => {
      const keyPath = path.join(tmpDir, 'custom_key.pem');
      const kp = generateKeypair();
      await fs.writeFile(keyPath, kp.privateKeyPem);

      const config = {
        jacsAgentName: 'test',
        jacsAgentVersion: '1.0',
        jacsKeyDir: tmpDir,
        jacsPrivateKeyPath: keyPath,
      };

      const pem = await loadPrivateKey(config);
      expect(pem).toContain('-----BEGIN PRIVATE KEY-----');
    });

    it('throws when jacsPrivateKeyPath is configured but missing', async () => {
      const keyPath = path.join(tmpDir, 'missing_key.pem');
      const kp = generateKeypair();
      await fs.writeFile(path.join(tmpDir, 'agent_private_key.pem'), kp.privateKeyPem);

      const config = {
        jacsAgentName: 'test',
        jacsAgentVersion: '1.0',
        jacsKeyDir: tmpDir,
        jacsPrivateKeyPath: keyPath,
      };

      await expect(loadPrivateKey(config)).rejects.toThrow(
        'Configured jacsPrivateKeyPath does not exist',
      );
    });

    it('strips comment lines from PEM', async () => {
      const keyDir = path.join(tmpDir, 'keys');
      await fs.mkdir(keyDir);

      const kp = generateKeypair();
      const commentedPem = '# WARNING: TEST ONLY\n' + kp.privateKeyPem;
      await fs.writeFile(path.join(keyDir, 'private_key.pem'), commentedPem);

      const config = {
        jacsAgentName: 'test',
        jacsAgentVersion: '1.0',
        jacsKeyDir: keyDir,
      };

      const pem = await loadPrivateKey(config);
      expect(pem).not.toContain('# WARNING');
      expect(pem).toContain('-----BEGIN PRIVATE KEY-----');
    });

    it('throws when no key found', async () => {
      const config = {
        jacsAgentName: 'test',
        jacsAgentVersion: '1.0',
        jacsKeyDir: tmpDir, // empty dir
      };

      await expect(loadPrivateKey(config)).rejects.toThrow('No private key found');
    });

    it('tries agent_private_key.pem as candidate', async () => {
      const keyDir = path.join(tmpDir, 'keys');
      await fs.mkdir(keyDir);

      const kp = generateKeypair();
      await fs.writeFile(path.join(keyDir, 'agent_private_key.pem'), kp.privateKeyPem);

      const config = {
        jacsAgentName: 'test',
        jacsAgentVersion: '1.0',
        jacsKeyDir: keyDir,
      };

      const pem = await loadPrivateKey(config);
      expect(pem).toContain('-----BEGIN PRIVATE KEY-----');
    });
  });

  describe('loadPrivateKeyPassphrase', () => {
    it('uses env password by default', async () => {
      process.env.JACS_PRIVATE_KEY_PASSWORD = 'dev-password';
      const passphrase = await loadPrivateKeyPassphrase();
      expect(passphrase).toBe('dev-password');
    });

    it('uses password file when env source is disabled', async () => {
      const passwordFile = path.join(tmpDir, 'password.txt');
      await fs.writeFile(passwordFile, 'file-password\n', 'utf-8');
      if (process.platform !== 'win32') {
        await fs.chmod(passwordFile, 0o600);
      }

      process.env.JACS_PRIVATE_KEY_PASSWORD = 'dev-password';
      process.env.JACS_PASSWORD_FILE = passwordFile;
      process.env.JACS_DISABLE_PASSWORD_ENV = '1';

      const passphrase = await loadPrivateKeyPassphrase();
      expect(passphrase).toBe('file-password');
    });

    it('throws when multiple password sources are configured', async () => {
      const passwordFile = path.join(tmpDir, 'password.txt');
      await fs.writeFile(passwordFile, 'file-password\n', 'utf-8');

      process.env.JACS_PRIVATE_KEY_PASSWORD = 'dev-password';
      process.env.JACS_PASSWORD_FILE = passwordFile;

      await expect(loadPrivateKeyPassphrase()).rejects.toThrow('Multiple password sources configured');
    });

    it('throws when no password source is configured', async () => {
      await expect(loadPrivateKeyPassphrase()).rejects.toThrow('Private key password required');
    });

    it('skips disabled file source and uses env', async () => {
      process.env.JACS_PRIVATE_KEY_PASSWORD = 'dev-password';
      process.env.JACS_PASSWORD_FILE = '/tmp/unused-password-file.txt';
      process.env.JACS_DISABLE_PASSWORD_FILE = 'true';

      const passphrase = await loadPrivateKeyPassphrase();
      expect(passphrase).toBe('dev-password');
    });

    it('rejects insecure password file permissions on unix', async () => {
      if (process.platform === 'win32') {
        return;
      }

      const passwordFile = path.join(tmpDir, 'password-insecure.txt');
      await fs.writeFile(passwordFile, 'file-password\n', 'utf-8');
      await fs.chmod(passwordFile, 0o644);

      process.env.JACS_PASSWORD_FILE = passwordFile;
      process.env.JACS_DISABLE_PASSWORD_ENV = '1';

      await expect(loadPrivateKeyPassphrase()).rejects.toThrow('insecure permissions');
    });

    it('rejects symlink password file sources', async () => {
      if (process.platform === 'win32') {
        return;
      }

      const targetFile = path.join(tmpDir, 'password-target.txt');
      const linkFile = path.join(tmpDir, 'password-link.txt');
      await fs.writeFile(targetFile, 'file-password\n', 'utf-8');
      await fs.chmod(targetFile, 0o600);
      await fs.symlink(targetFile, linkFile);

      process.env.JACS_PASSWORD_FILE = linkFile;
      process.env.JACS_DISABLE_PASSWORD_ENV = '1';

      await expect(loadPrivateKeyPassphrase()).rejects.toThrow('must not be a symlink');
    });
  });
});
